pkg/controllers/workspace_controller.go

```diff
@@ -56,40 +59,53 @@ func (c *WorkspaceReconciler) Reconcile(ctx context.Context, req reconcile.Reque
 func (c *WorkspaceReconciler) applyWorkspaceResource(ctx context.Context, wObj *kdmv1alpha1.Workspace) error {
 	klog.InfoS("applyWorkspaceResource", "workspace", klog.KObj(wObj))
 	// Check CandidateNodes, if the count, instance type
-	candidateNodeCount, err := c.validateCandidateNodes(ctx, wObj)
+	validNodeList, err := c.validateCandidateNodes(ctx, wObj)
 	if err != nil {
 		return err
 	}
 
+	candidateNodeCount := len(validNodeList)
 	remainingNodeCount := lo.FromPtr(wObj.Resource.Count) - candidateNodeCount
+
 	// if current node Count == workspace count, then all good and return
 	if remainingNodeCount <= 0 {
-		klog.InfoS("number of candidate nodes, %d, is equal or greater than required workspace count, %d ",
-			candidateNodeCount, lo.FromPtr(wObj.Resource.Count))
+		klog.InfoS("number of candidate nodes, is equal or greater than required workspace count", "workspace.Count", lo.FromPtr(wObj.Resource.Count))
 		return nil
-	}
-	klog.InfoS("need to create more nodes", "remainingNodeCount", remainingNodeCount)
-	klog.InfoS("Will start building a machine")
+	} else {
+		klog.InfoS("need to create more nodes", "remainingNodeCount", remainingNodeCount)
+		klog.InfoS("Will start building a machine")
+		// TODO Create machine CR
 
-	// Create machine CR
+	}
 
-	// Add nodes names to the WorkspaceStatus.WorkerNodes[]
+	// Ensure all node plugins are running successfully
+	for i := range validNodeList {
+		err = node.EnsureNodePlugins(ctx, validNodeList[i], c.Client)
+		if err != nil {
+			return err
+		}
+	}
+	// Add the valid nodes names to the WorkspaceStatus.WorkerNodes
+	err = c.updateWorkspaceStatusWithNodeList(ctx, wObj, validNodeList)
+	if err != nil {
+		return err
+	}
 
 	return nil
 }
 
-func (c *WorkspaceReconciler) validateCandidateNodes(ctx context.Context, wObj *kdmv1alpha1.Workspace) (int, error) {
+func (c *WorkspaceReconciler) validateCandidateNodes(ctx context.Context, wObj *kdmv1alpha1.Workspace) ([]*corev1.Node, error) {
 	klog.InfoS("validateCandidateNodes", "workspace", klog.KObj(wObj))
-	validNodesCount := 0
+	var validNodeList []*corev1.Node
 	nodeList := wObj.Resource.CandidateNodes
```

Comment:
I think in validateCandidateNodes, you can scan all hosts. 


Later, you should have a filter function to find out the existing candidates by considering spec.resource.CandidateNotes, all existing candidate nodes, and the nodelist saved in the status.workers.

Reply:
Addressed.